### ISSUE TITLE
NPE using mutiny cursor - possibly due to number of fields

### ISSUE THREAD

**Issue by Biratus** [quarkusio/quarkus#10114]
Hi, I am using a cursor to go through large amounts of data. I implemented the cursor using Mutiny's Uni and Multi.
It works great but once compiled in the native image I get a NPE out of nowhere that is not catch by the onFailure. I think the NPE closes the database connection and then onFailure catched the SQLException: closed connection.
The NPE is thrown no matter the cursor's read size or the number of data to retrieve. It works with 0 data to select though..

I have made a reproducer and identified the error. It seems the number of fields retreived from the database is the cause. By commenting some lines you should be able to reproduce the error (see README)

I've opened this thread on zulip [https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/SQL.20Cursor.20-.20Mutiny.20-.20Native.20image](https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/SQL.20Cursor.20-.20Mutiny.20-.20Native.20image )
And you can find the reproducer here: [https://github.com/Biratus/test-cursor-reproducer](https://github.com/Biratus/test-cursor-reproducer)

In the Zulip thread Thomas suggest another approach to using the cursor. Both are implemented in the reproducer and both have the same behaviour. 

The error happens with quarkus 1.4.1.Final and 1.5.1.Final

The code is unecessary complex, it is to mirror the original module on which the error occured.

**Comment by Biratus** [comment]
up

**Comment by cescoffier** [comment]
cc @tsegismont

**Comment by tsegismont** [comment]
I'll look into it

**Comment by tsegismont** [comment]
@Biratus can you please add to the reproducer README the commands to run in order to:

- start the DB with Docker and insert test data
- compile, package and run the app
- send a request to generate the failure

Thank you

**Comment by tsegismont** [comment]
So without precise instructions for reproducing I went ahead with:

```
docker run -p3306:3306 -e MYSQL_ROOT_PASSWORD=password -e MYSQL_DATABASE=rv_relai_aa -e MYSQL_USER=rv_relai_aa -e MYSQL_PASSWORD=password -it mysql
```

My `native-image` version:

```
 ${GRAALVM_HOME}/bin/native-image --version
GraalVM Version 19.3.1 CE
```

I first started the program in JVM mode with `mvn clean quarkus:dev`
Then browsed to `localhost:8080/mysql` and got the list of rows in the console.

I couldn't compile to native so I had to modify the POM:

```patch
Index: pom.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- pom.xml	(revision 146147d235506b8e43efadf1643fce7ff766a31c)
+++ pom.xml	(date 1595425644537)
@@ -17,10 +17,6 @@
     <quarkus.platform.group-id>io.quarkus</quarkus.platform.group-id>
     <quarkus.platform.version>1.5.1.Final</quarkus.platform.version>
     <surefire-plugin.version>2.22.1</surefire-plugin.version>
-    <native.command.default>
-      --no-server,-Dio.netty.noUnsafe=true,-Dgraal.CompilationFailureAction=Diagnose,-H:ReflectionConfigurationFiles=${project.basedir}/reflectconfigs/config.json,-Djava.util.logging.manager=org.jboss.logmanager.LogManager,--report-unsupported-elements-at-runtime
-    </native.command.default>
-    <native.command>${native.command.default}</native.command>
   </properties>
   <dependencyManagement>
     <dependencies>
@@ -127,7 +123,6 @@
                   <goal>native-image</goal>
                 </goals>
                 <configuration>
-                  <additionalBuildArgs>${native.command},-H:+TraceClassInitialization</additionalBuildArgs>
                   <enableHttpUrlHandler>true</enableHttpUrlHandler>
                   <enableAllSecurityServices>true</enableAllSecurityServices>
                   <enableHttpsUrlHandler>true</enableHttpsUrlHandler>
```

Then ran the generated `runner` program.
When I browsed to `localhost:8080/mysql` again, I also got the list of rows in the console.

@Biratus without more info I can't do much.

cc @cescoffier

**Comment by cescoffier** [comment]
Thanks @tsegismont! 

@Biratus please provide the instructions. Your bug may have been fixed in between, so please check if it's still valid on the last Quarkus version (1.6.1).

**Comment by benDeMtp** [comment]
Hello, 
@Biratus has left the project.
So I just checkout his reproducer and I have reproduced, without any change to the pom.

`mvn install`
then 
`java -jar target/test-cursor-mysql-1.0-SNAPSHOT-runner.jar`

Goto http://127.0.0.1:8080/mysql and see the logs. Test OK

For native image

`mvn package -Pnative`
then
`target/test-cursor-mysql-1.0-SNAPSHOT-runner`

Go to  http://127.0.0.1:8080/mysql and see the logs. Test is KO with this logs:

`java.lang.NullPointerException
        at io.netty.util.internal.PlatformDependent0.throwException(PlatformDependent0.java:450)
        at io.netty.util.internal.CleanerJava6.freeDirectBuffer(CleanerJava6.java:114)
        at io.netty.util.internal.PlatformDependent.freeDirectBuffer(PlatformDependent.java:479)
        at io.netty.buffer.PoolArena$DirectArena.destroyChunk(PoolArena.java:763)
        at io.netty.buffer.PoolArena.free(PoolArena.java:271)
        at io.netty.buffer.PooledByteBuf.deallocate(PooledByteBuf.java:171)
        at io.netty.buffer.AbstractReferenceCountedByteBuf.handleRelease(AbstractReferenceCountedByteBuf.java:110)
        at io.netty.buffer.AbstractReferenceCountedByteBuf.release(AbstractReferenceCountedByteBuf.java:100)
        at io.netty.handler.codec.ByteToMessageDecoder$1.cumulate(ByteToMessageDecoder.java:105)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:274)
        at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:527)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:193)
`

@tsegismont for mysql we use this configuration in compose file:
`version: '3.7'
services:

  mysql:
    container_name: mysql
    image: mysql:5.7
    environment:
      - MYSQL_ROOT_PASSWORD=password
      - MYSQL_DATABASE=rv_relai_aa
      - MYSQL_USER=rv_relai_aa
      - MYSQL_PASSWORD=password
      - TZ=Europe/Paris
    ports:
      - "3306:3306"
`

@cescoffier and @tsegismont here is the output of java -version:
`openjdk version "1.8.0_242"
OpenJDK Runtime Environment (build 1.8.0_242-b06)
OpenJDK 64-Bit Server VM GraalVM CE 20.0.0 (build 25.242-b06-jvmci-20.0-b02, mixed mode)
`

@cescoffier does this version compliant with 1.6 ? 

Thanks a lot

**Comment by benDeMtp** [comment]
I just tried with 1.6.1 and this java version : openjdk version "11.0.7" 2020-04-14
OpenJDK Runtime Environment GraalVM CE 20.1.0 (build 11.0.7+10-jvmci-20.1-b02)
OpenJDK 64-Bit Server VM GraalVM CE 20.1.0 (build 11.0.7+10-jvmci-20.1-b02, mixed mode, sharing)

And it works like @tsegismont said.
I presume it has been fixed in 1.6.x

I'll try in the real project now :)

**Comment by geoand** [comment]
Closing as this has pressumably been fixed. 
If not, feel free to reopen
